Use TabPanel in edit settings sidebar header (#13587)

jffng opened the pull request:
## Description
This PR replaces the sidebar settings header with `TabPanel` from the components package. 

No visual change, but a slight behavioral change in the keyboard navigation: only the active tab is placed in the tab order, and the user can navigate between tabs using the arrows. 

This follows the [WAI-ARIA recommendation](https://www.w3.org/TR/wai-aria-practices-1.1/) on keyboard interaction for tab panel components:

> Tab: When focus moves into the tab list, places focus on the active tab element . When the tab list contains the focus, moves focus to the next element in the page tab sequence outside the tablist, which is typically either the first focusable element inside the tab panel or the tab panel itself.

## How has this been tested?
Tested in the playground and editor on:

macOS 10.14.5
Chrome 75.0.3770.100
Safari 12.1.1
Firefox 67.0.4

## Types of changes
Addresses #13587

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

jffng committed:
Use TabPanel in edit settings sidebar header (#13587)

* Add styles to the component itself

* Add aria-label prop to the TabButton

* Replace markup with TabPanel component in edit settings sidebar header

jffng commented:
Discovered two issues to address: 

1. Rework some of the tests as the class selectors have changed.
2. Active tab class needs to change when the `edit-post/block` event fires

jffng committed:
Toggle active states when block is selected
Fix style border bug

jffng committed:
Refactor TabPanel with useState and useEffect

Modify e2e tests to use arrow keys to switch between tabs and select using
aria-label

Allow settings header to control the active tab

ItsJonQ commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -3,54 +3,84 @@
+				onSelect={ onSelect }
+				controlledTabName={ selectedTabName }
+			>
+				{ () => { } }
```
@jffng Hallooo! This PR looks great! I was a little confused at this one (at first glance).

What do you think about adjusted the original `TabPanel` component's `children()` render to do something like...

`children && children(selectedTab)`

That way, you won't have to pass it in a children render function that returns null/empty.

P.S. I'm not sure what the current conventions are for handling this flow 😊. The flow I suggested feels little cleaner + less ambiguous

jffng replied:
Thanks @ItsJonQ! I agree. We could also set `children`'s default value to `noop`.

ItsJonQ replied:
Aww yiss! `noop` it up! 🎉

mapk commented:
Thanks @jffng for a fantastic PR, and thanks @ItsJonQ for the follow-up code review. I just tested this and love jumping around with the arrow keys. I did notice some keyboard nav weirdness though. When I was in the tabPanel, and "Document" was focused, hitting <tab> brought me to an in-between state when clicking <Enter> didn't do anything and nothing was focused. If I hit <tab> again, I was taken to the "Close" icon. Is this expected behavior?

I feel like that in-between tab position _wants_ to rest on the "Block" tab, but isn't so it just leaves me hanging.

Would love an accessibility review here too!

jffng committed:
Backwards compatibility for initialTabName prop

jffng commented on packages/components/src/tab-panel/README.md:
```diff
@@ -133,9 +133,9 @@ The class to add to the active tab
 - Default: `is-active`
 
-#### initialTabName
+#### controlledTabName
```
Thanks for the review @gziolo, noted. 

I added support for the `initialTabName` for backwards compatibility, as well deprecation notes in the README. However, I lack context as to the timing / version to officially deprecate this prop. 

This is also assuming that having a prop to control the active tab from a parent component is a good idea / not an anti-pattern.

gziolo replied:
If we don't want to keep this param, we should rather remove it from docs and ensure that when used it works as close as possible as the new one.

jffng committed:
Don't render the tab panel content container unless there's content

Adjust test to three tab presses to navigate the block settings instead of four

jffng commented:
> When I was in the tabPanel, and "Document" was focused, hitting `tab` brought me to an in-between state where clicking `Enter` didn't do anything and nothing was focused. If I hit `tab` again, I was taken to the "Close" icon. Is this expected behavior?

No, nice catch @mapk. The component was rendering an empty container where the tab panel content would normally go (we're not actually using it in this case), and this empty container was receiving the focus in that in-between state. I pushed a fix that should resolve it, making the tab panel content optional.

> Would love an accessibility review here too!

Yes!

gziolo commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -3,54 +3,84 @@
+		}
+	};
+
+	useEffect(
```
Did you consider adding `onSelect` to the objects passed to `tabs` instead? It might simplify the solution as you could leave the whole logic which reasons which tab is selected to `TabPanel` component.

jffng replied:
I had considered it, but avoided that route as I was originally trying to minimize changes to `TabPanel`. I agree adding an `onSelect` method to the tab objects would simplify the solution. (We would still need to support the `onSelect` component handler for backwards compatibility)

The issue that led me to refactor `TabPanel` was the need to control the selected tab from the parent component without triggering a re-render of the entire component, causing the active tab to lose focus.

gziolo replied:
> The issue that led me to refactor `TabPanel` was the need to control the selected tab from the parent component without triggering a re-render of the entire component, causing the active tab to lose focus.

Shouldn't the component ensure that there is never focus loss? 3rd party developers shouldn't care about it, it should be baked in into such component.

jffng replied:
I will think about that, thank you!

jffng replied:
I believe I was able to simplify the logic, please let me know if [these changes](https://github.com/WordPress/gutenberg/pull/16663/commits/068604458b3fe690d7d041011850056104a0ec30) address your feedback @gziolo, thanks.

jffng committed:
Simplify settings header and tab panel logic

gziolo reviewed (commented):
@jffng, the changes applied to PR look great. This is exactly what I would see as someone who consumes this component. Everything is handled behind the scenes and you don't need to worry about all the logic which handles navigation between tabs.

jffng committed:
Differ aria-label selected text to TabPanel

jffng commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -3,54 +3,53 @@
+	const blockAriaLabel = sidebarName === 'edit-post/block' ?
 		// translators: ARIA label for the Settings Sidebar tab, selected.
-		[ __( 'Block (selected)' ), 'is-active' ] :
+		__( 'Block (selected)' ) :
```
Nice suggestion, this simplifies the settings header logic even further.

gziolo commented:
@drw158 - how do you feel about the changes proposed in the context of the audit that you performed and documented in #16710?

As far as I see, `TabPanel` is used for the content of the tab in other libraries. I checked myself these two libs:
- https://reakit.io/docs/tab/
- https://ui.reach.tech/tabs

drw158 commented:
@gziolo I'm leaning towards either structure as shown in Reakit or Reach UI, especially if we rename our current Panel component.

```
<TabList {...tab} aria-label="My tabs">
        <Tab {...tab} stopId="tab1">
          Tab 1
        </Tab>
        <Tab {...tab} stopId="tab2" disabled>
          Tab 2
        </Tab>
        <Tab {...tab} stopId="tab3">
          Tab 3
        </Tab>
      </TabList>
      <TabPanel {...tab} stopId="tab1">
        Tab 1
      </TabPanel>
      <TabPanel {...tab} stopId="tab2">
        Tab 2
      </TabPanel>
      <TabPanel {...tab} stopId="tab3">
        Tab 3
      </TabPanel>
```

annezazu commented:
It seems the original issue was closed https://github.com/WordPress/gutenberg/issues/13587 and this problem was addressed in another PR: https://github.com/WordPress/gutenberg/pull/20872 Going to close this out as a result :) Feel free to re-open/ask for it to be if I am missing something.
